XRT-3292: Dockerised XRT Examples

samtomcom opened the pull request:
https://iotechsys.atlassian.net/browse/XRT-3292

Dockerised example, with multiple devices, opc-ua, and grafana dashboard containers.

JonathanWhiteUK committed:
XRT-3292: Build on initial work from Jonathan

- Stripped down to just devices to start adding components back

samtomcom committed:
XRT-3292: Added OPC-UA server/browser

- Added command component to XRT instances to allow federated deployment

samtomcom committed:
XRT-3292: Store common variables in .env file

- Replace service_healthy with service_started

samtomcom committed:
XRT-3929: Add grafana dashboards and influxdb xrt instance.

- Make other instances of xrt have unique names to avoid clashes

samtomcom committed:
XRT-3292: Remove mqtt dependancies so that it can be stopped/started independantly

samtomcom committed:
XRT-3292: Add README, and remove temporary ideas.zip

samtomcom committed:
XRT-3292: More detailed README

samtomcom committed:
XRT-3292: Remove reference to SKF dashboard

samtomcom committed:
Change bacnet example schedule to all

BinduRao2018 commented on DeviceServices/s7/deployment/config/mqtt_bridge.json:
```diff
@@ -2,15 +2,21 @@
-  "MQTTPatterns": [ "xrt/devices/s7/request" ],
+  "MQTTPatterns": [
+    "xrt/request",
+    "xrt/discovery",
```
No Discovery on S7?

samtomcom replied:
I only added new patterns, didn't remove any existing ones for minimal functionality. Is `xrt/devices/s7/discovery` needed for something in this example?

BinduRao2018 commented on Dockerised/README.md:
```diff
@@ -0,0 +1,104 @@
+COMPOSE_PROFILES=* docker compose down
+
+# Alternatively
+docker compose --profiles "*" up -d
```
--profiles option doesn't work with Docker Compose version 2.29

samtomcom replied:
Ah it's just `--profile` without the `s`, will fix

BinduRao2018 commented on Dockerised/.env:
```diff
@@ -0,0 +1,19 @@
+
+# For xrt env
+XRT_CONFIG_DIR="/opt/iotech/xrt/2.2/deployment/config"
+XRT_LICENSE_FILE="/opt/iotech/xrt/2.2/IOTech_EdgeConnect_Developer.json"
```
The license file cannot be shared. 
It is ok to add this example within the Xrt repository with the license file, but not in xrt-examples, which is a public repository

samtomcom replied:
This is just the env var for containers to use internally to find the license file inside the container, which is mounted as a volume using the users local license. Not actually sharing any new licenses anywhere.

BinduRao2018 commented on Dockerised/deployment/deployment/config/influxdb_exporter.json:
```diff
@@ -0,0 +1,25 @@
+{
```
Why is this deployment copied, instead of referring to xrt-examples?

samtomcom replied:
Existing `Exporter/influxdb` example also runs another virtual device service which detracts from the purpose of the dockerised examples.

samtomcom replied:
It would be theoretically possible, but would require changing the existing `Exporter/influxdb` which may make it not as useful as it's own example - but up to you.

BinduRao2018 commented on Dockerised/docker-compose.yml:
```diff
@@ -0,0 +1,204 @@
+  influxdb-data:
+
+services:
+  mqtt:
```
Use the same version of mqtt and influxdb used for testing Xrt

JonathanWhiteUK replied:
Gotten Sam to fix the influxdb version to 2.x.x as shouldn't be any breaking changes in that for us.

I don't think the mqtt brokers version needs to be fixed as the MQTT Protocol is stable and not changing. In the same way you could swap this container for EMQX or NANOMQ and still expect it to work.

BinduRao2018 commented on Dockerised/virtual/mosquitto.conf:
```diff
@@ -0,0 +1,30 @@
+# Set the listener port for MQTT connections
```
Is this specific configuration for the MQTT Broker required?

samtomcom replied:
I will try it without it.

samtomcom replied:
Yes it's needed, because the MQTT is running in it's own container it needs this config. Each xrt instance has it's own .conf file when it's run normally, but with MQTT in it's own container it needs to be provided.

BinduRao2018 reviewed (dismissed):
- Toplevel Readme to be updated to add a link to this Dockerised example
- For the DS that doesn't support discovery or Events, delete "xrt/discovery", "xrt/<>/event topic configuration.

samtomcom committed:
Address review comments

* top level readme
* s/profiles/profile
* inform about license file
* set influxdb version
* remove uncessary discovery/event topics

samtomcom committed:
Revert removing discovery/event topics for OPC-UA to see devices

JonathanWhiteUK commented on Dockerised/README.md:
````diff
@@ -0,0 +1,113 @@
+../DeviceServices/virtual/commands/get_request.sh
+```
+
+Note that some devices/resources/schedules information in the `commands/` scripts may be specific to those examples and want changing to see any impact on the running dockerised examples.
````
review

samtomcom replied:
Changed device command to modbus-tcp and removed comment about changing scripts to prevent confusion.

JonathanWhiteUK commented on Dockerised/README.md:
```diff
@@ -0,0 +1,113 @@
+This dashboard will show saved data from devices running and saved to InfluxDB by another containerised instance of XRT.
+
+The Grafana config files are held in `./deployment/grafana`.
+The Exporter XRT instance config files are held in `./deployment/deployment`.
```
review

samtomcom replied:
Renamed directory to `deployment/exporter`

JonathanWhiteUK commented on Dockerised/deployment/grafana/dashboards/main.yaml:
```diff
@@ -0,0 +1,12 @@
+apiVersion: 1
```
why do we have 2 of these

samtomcom replied:
Deleted duplicate

JonathanWhiteUK commented on Dockerised/docker-compose.yml:
```diff
@@ -0,0 +1,204 @@
+
+  opc-ua-browser:
+    container_name: xrt-opc-ua-browser
+    image: iotechsys/opc-ua-browser:1.1.0
```
use latest

JonathanWhiteUK commented on Dockerised/docker-compose.yml:
```diff
@@ -0,0 +1,204 @@
+    image: eclipse-mosquitto:latest
+    volumes:
+      - ./virtual/mosquitto.conf:/mosquitto/config/mosquitto.conf
+    ports:
```
souble check if mqtt need a profile

samtomcom replied:
Looks like it's needed, even with virtual and s7 both having `depends_on: mqtt`, killing one of the devices leaves the other running and mqtt killed

```
sam@samc ~/code/xrt-examples/Dockerised
 % docker ps
CONTAINER ID   IMAGE     COMMAND   CREATED   STATUS    PORTS     NAMES
sam@samc ~/code/xrt-examples/Dockerised
 % COMPOSE_PROFILES=virtual,s7 docker compose up -d
[+] Running 5/5
 ✔ Network dockerised_iotech    Created                                                                                                                                                  0.1s 
 ✔ Container dockerised-mqtt-1  Started                                                                                                                                                  0.2s 
 ✔ Container s7-simulator       Started                                                                                                                                                  0.2s 
 ✔ Container xrt-virtual-ds     Started                                                                                                                                                  0.3s 
 ✔ Container xrt-s7-ds          Started                                                                                                                                                  0.3s 
sam@samc ~/code/xrt-examples/Dockerised
 % COMPOSE_PROFILES=virtual docker compose down    
[+] Running 3/3
 ✔ Container xrt-virtual-ds     Removed                                                                                                                                                  0.9s 
 ✔ Container dockerised-mqtt-1  Removed                                                                                                                                                  0.3s 
 ! Network dockerised_iotech    Resource is still in use                                                                                                                                 0.0s 
sam@samc ~/code/xrt-examples/Dockerised
 % docker ps
CONTAINER ID   IMAGE                  COMMAND                  CREATED          STATUS          PORTS     NAMES
c84836dcecec   iotechsys/xrt:2.2      "sh -c '$XRT/bin/xrt…"   11 seconds ago   Up 9 seconds              xrt-s7-ds
4a78dae0527a   iotechsys/s7-sim:2.0   "./s7_server_app"        11 seconds ago   Up 10 seconds             s7-simulator
```

samtomcom committed:
Address review comments 2

remove device commands
rename deployment/deployment to deployment/exporter
delete duplicate dashboard file in grafana config
use latest opc ua browser container
undo changes to federated example

samtomcom committed:
Remove mqtt specific profile

samtomcom committed:
fixup

samtomcom committed:
fixup: depend on mqtt
